### PR TITLE
Pagination refactoring

### LINKED ISSUE #12: Refactor methods for pagination
The code from #1 contains a lot of copy & paste code for all the pagination loops.
This issue is to try to refactor it into something cleaner and generic.

Also, the pagination can be improved to be aware of how many elements we already retrieved, and how many are left.
For example, for a repository with 140 issues, maybe we can have a good default for the first page of 20. The `totalCount` returns 140, so we know we can do another query of 100 and another one of 20.

### PR DESCRIPTION
**Pull request by smacker**
Fix: #12

This PR removes duplicated code and reduces the number of hard-coded string in the code.
In terms of lines diff:
```
2 files changed, 393 insertions(+), 614 deletions(-)
```
But more importantly, it's harder to make a mistake as downloading connections is unified.

---

* [ ] I have updated the CHANGELOG file according to the conventions in [keepachangelog.com](https://keepachangelog.com)
* [x] This PR contains changes that do not require a mention in the CHANGELOG file

### CODE REVIEW

**Comment by smacker** [comment]
rebased on master

**Review by se7entyse7en — CHANGES_REQUESTED** [review]
some little changes 👏

File: github/downloader.go [hunk 1]
```diff
@@ -250,688 +225,499 @@ func (d Downloader) RateRemaining(ctx context.Context) (int, error) {
...
+	ctx context.Context,
+	t connectionType,
+	res Connection,
+	q Query,
+	variables map[string]interface{},
+	process func(Connection) error,
+) error {
+	logger := ctxlog.Get(ctx)
+	// logging only top-level resources
+	isLoggable := t == issuesType || t == pullRequestsType || t == membersWithRole
```
[the remark this answers is not in the document]
Comment (smacker) [reply github/downloader.go]: Is this ok? https://github.com/src-d/metadata-retrieval/pull/41/commits/732fa1325870279de47daddc4f04fcb03bfcc692

File: github/downloader.go [hunk 2]
```diff
@@ -250,688 +225,499 @@ func (d Downloader) RateRemaining(ctx context.Context) (int, error) {
...
+		variables[t.Cursor()] = githubv4.String(res.GetPageInfo().EndCursor)
 
-		// get only repository topics
-		var q struct {
-			Node struct {
-				Repository struct {
-					RepositoryTopics graphql.RepositoryTopicsConnection `graphql:"repositoryTopics(first: $repositoryTopicsPage, after: $repositoryTopicsCursor)"`
-				} `graphql:"... on Repository"`
-			} `graphql:"node(id:$id)"`
+		if isLoggable && count%int(t.PageSize) == 0 {
```
[the remark this answers is not in the document]
Comment (smacker) [reply github/downloader.go]: I made it log each page. As now we use max limit 100, it should be fine.

**Review by se7entyse7en — APPROVED** [review]
Thanks!

**Pull request merged**